src/twisted/words/protocols/jabber/xmlstream.py

```diff
@@ -400,21 +401,39 @@ class TLSInitiatingInitializer(BaseFeatureInitiatingInitializer):
     set the C{wanted} attribute to False instead of removing it from the list
     of initializers, so a proper exception L{TLSRequired} can be raised.
 
-    @cvar wanted: indicates if TLS negotiation is wanted.
+    @ivar wanted: indicates if TLS negotiation is wanted.
     @type wanted: C{bool}
+
+    @ivar contextFactory: An object which creates appropriately configured TLS
+        connections. This is passed to C{startTLS} on the transport and is
+        preferably created using L{twisted.internet.ssl.optionsForClientTLS}.
+        If C{None}, the default is to verify the server certificate against
+        the trust roots as provided by the platform. See
+        L{twisted.internet._sslverify.platformTrust}.
+    @type contextFactory: L{IOpenSSLClientConnectionCreator}
     """
 
     feature = (NS_XMPP_TLS, 'starttls')
     wanted = True
+    contextFactory = None
     _deferred = None
 
+    def __init__(self, xs, required=True, contextFactory=None):
+        super(TLSInitiatingInitializer, self).__init__(
+                xs, required=required)
+        self.contextFactory = contextFactory
+
+
     def onProceed(self, obj):
         """
         Proceed with TLS negotiation and reset the XML stream.
         """
 
         self.xmlstream.removeObserver('/failure', self.onFailure)
-        ctx = ssl.CertificateOptions()
+        if self.contextFactory:
```

Comment:
Any chance this conditional could be done in the initializer so that `contextFactory` could be always set to something other than `None` on instances?  Attributes which can randomly change their type around are a magnet for bugs.

Reply:
I tried this, and this worked, but only if pyOpenSSL is available. To get no tests to fail, I'd have to skip quite a few more, as constructing the initializer will now fail if OpenSSL is not available, even if the client or server don't require TLS. I think it constitutes a breaking change.

Reply:
I reverted this particular change.